Enhance filtering and updating for set method

sygutss opened the pull request:
1. Added possibility to filter not only leaves but all dict structures like in search.
Flag parameter "only_leaves" - by default is True and functionality is off (Backward compatibility)

2. Added possibility to update dictionaries (works only with "only_leaves" set to False
and dictionaries nodes). With this flag we can update specified node (dictionary).
By default is set to False (Backward compatibility)

moomoohk commented:
@sygutss Hey, are you still interested in working on this PR?

sygutss commented:
Why not :)  I can have a look in the meantime

moomoohk commented:
Awesome! If you have any questions feel free to reach out here or on Gitter.

sygutss commented:
I've adjusted to the latest version of the master branch and refactored one method '_split_path' to be more pythonic
and concise :)

moomoohk commented:
Could you please provide some example code to give me a better idea of what this PR proposes?

sygutss commented:
1. Found a small bug in my implementation and I've fixed it in the previous commit - right now it should work as expected

2. I've added tests for that and in the previous commit I've changed a bit them - maybe right now would be clearer what's the purpose of that functionality - Treat them as an example :) 

Description:
The main purpose of my enhancement was to provide a possibility to filter not only via values from the last level - the leaves level (basic types under keys) but also from the higher levels - so using entire objects (dictionaries) from whichever level. 

With that approach we can eg. indicate some known levels Eg. "A/B/*" - in that case, are 'A' and 'B' known but the rest is unknown so we use "*". What else, we know that on some level there should be a specific key with some specific value eg. some type. We can filter using a provided function that is able to compare a specific key from the object from some level.

Additionally, I've added the dictionary update possibility instead of the entire replacement during the setting process.
